### PR TITLE
Set user environment variables permanently by using `--permanent` + deprecate `--global`

### LINKED ISSUE #189: emsdk construct_env on Windows nukes local paths
Windows-only repro: 
* open a console
* add something to the path: `> set path=%path%;something`
* `> emsdk_env`
* Your something is gone.

(conversely remove something and see it reappear)

The problem is that emsdk gets the path to which it appends from the registry so any local changes are lost. I was tempted to make a pull request which fixes this, but since specific effort is made for it to work like this, I decided to ask first.

The relevant code is here: [emsdk#L1935](https://github.com/juj/emsdk/blob/7a0e27441eda6cb0e3f1210e6837cae4b080ab4c/emsdk#L1935)

Why get the system path in such convoluted way? Why query the registry instead of just use the path environment variable?

### LINKED ISSUE #467: Way to set user environment variables
Currently `tool.win_activate_env_vars` always gets `permanently_activate=True`, which then leads to setting global environment variables. I need a way to set it for local user scope.

### PR DESCRIPTION
**Pull request by aminya**
Closes https://github.com/emscripten-core/emsdk/pull/468

Fixes #467

This makes ./emsdk activate --global to set user environment variables by default instead of previous system-wide-only activation.

Also removes some magic that implicitly nukes user envvars when system ones are shadowed, as a system admin should understand what's going on when setting envvars. Thus fixes #189 and closes #192.

### CODE REVIEW

**Comment by sbc100** [comment]
According to the comments in #468 we decided it was best *not* to change existing meaning of `--global`.

I think the best option was to add two new options `--persistent` (meaning set in the registry for the current user) and `--persistent-system` meaning set in the registry for all users.  

`--global` can the becomes a legacy alias for `--persistent-system`.

**Comment by aminya** [comment]
I deprecated `--global`. Now, we have `--permanent` to set the environment variables permanently.

**Comment by aminya** [comment]
I deprecated `--global`. Now, we have `--permanent` to set the environment variables permanently.

**Comment by aminya** [comment]
I added the tests, which shows that pywin32 code part of emsdk is broken.

https://github.com/emscripten-core/emsdk/pull/91

**Comment by saschanaz** [comment]
Yay, thanks for taking over!

File: emsdk.py [hunk 1]
```diff
@@ -2775,7 +2770,14 @@ def extract_bool_arg(name):
 
   arg_old = extract_bool_arg('--old')
   arg_uses = extract_bool_arg('--uses')
+  arg_permanent = extract_bool_arg('--permanent')
```
Comment (sbc100) [inline emsdk.py:2773]: Could you use persistent as discussed in the other bug?  I know that permanent and persistent are essentially synonyms but my sense is that persistent is more commonly used the tech world.

File: emsdk.py [hunk 2]
```diff
@@ -2775,7 +2770,14 @@ def extract_bool_arg(name):
...
   arg_global = extract_bool_arg('--global')
+  if arg_global:
+    print('--global is deprecated. Use `--system` to set the environment variables for all users')
+  arg_system = extract_bool_arg('--system')
+  if arg_system:
+    arg_permanent = True
+    arg_global = True
```
Comment (sbc100) [inline emsdk.py:2780]: Did you mean to set arg_global here?  I would imagine `arg_global` should no longer be used anywhere after this change?

File: emsdk.py [hunk 3]
```diff
@@ -2775,7 +2770,14 @@ def extract_bool_arg(name):
 
   arg_old = extract_bool_arg('--old')
   arg_uses = extract_bool_arg('--uses')
+  arg_permanent = extract_bool_arg('--permanent')
   arg_global = extract_bool_arg('--global')
+  if arg_global:
+    print('--global is deprecated. Use `--system` to set the environment variables for all users')
```
Comment (sbc100) [inline emsdk.py:2776]: Why not print the warning but continue to do the right thing and set arg_system here?  We can phase it out eventually.

File: scripts/test_permanent.ps1 [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    $EMSDK_UPSTREAM_Path_USER = $path_split | Where-Object { $_ -match 'emsdk\\upstream\\emscripten' }
+    if (!$EMSDK_UPSTREAM_Path_USER) {
+        throw "emsdk\upstream\emscripten is not added to path."
+    }
+
+}
+finally
+{
+# return back to cwd
+cd $cwd
```
Comment (sbc100) [inline scripts/test_permanent.ps1:66]: If you don't do this then does the directory change effect the calling shell?

**Review by sbc100 — COMMENTED** [review]
I've not had a chance to look at the new test script but generally this change lgtm!

File: emsdk.py (hunk 1 above)
Comment (aminya) [reply emsdk.py:2773]: That is the term that has been used in the readme: https://github.com/emscripten-core/emsdk#how-do-i-change-the-currently-active-sdk-version

I don't mind changing it, but I found permanent more natural.

File: emsdk.py (hunk 3 above)
Comment (aminya) [reply emsdk.py:2776]: Fixed in https://github.com/emscripten-core/emsdk/pull/621/commits/25efaf464a58404ecc72d39342a375df6a43f866

File: emsdk.py (hunk 2 above)
Comment (aminya) [reply emsdk.py]: Fixed in https://github.com/emscripten-core/emsdk/pull/621/commits/bd028b9b6278e6e7b58b972c11de62ce3b0e3748

File: scripts/test_permanent.ps1 (hunk 4 above)
Comment (aminya) [reply scripts/test_permanent.ps1:66]: This is so if the program encounters an error or the user press CTRL+C, the current working directory will be restored to where it was. This is a safe way to change the current working directory.

Comment (sbc100) [reply scripts/test_permanent.ps1:66]: Just to be clear, it sounds like the `cd` command effect *does* effect the calling process?   e.g. if I run `foo.sp1` can it change the current working directory in my shell or are the changes limited to the duration of the `foo.ps1` script?

Comment (aminya) [reply scripts/test_permanent.ps1:66]: Yes. It does. It changes the current working directory of the parent process. That's why I need to use `finally`.

Comment (aminya) [reply scripts/test_permanent.ps1:66]: Another way is to use `Push-Location` in `try` and then call `Pop-Location` in the `finally`. It is the same as my method.

**Comment by kripken** [comment]
Is the CI error the last issue here? (maybe also the permanent/persistent naming question)

**Comment by sbc100** [comment]
I think we need a tie breaker :)   @kripken @juj how to do you feel about `--persistent` vs `--permanent` . 

As a non-windows-user I don't care too much either way.

**Comment by aminya** [comment]
> Is the CI error the last issue here?

This is not an issue of this PR. I added the tests, so we know that environment variables are not set on Windows (the behavior is broken). We need to fix the tests in other pull requests such as in #91.

> (maybe also the permanent/persistent naming question)

I don't feel strongly about this. I think `--permanent` is more descriptive. `--persistent` is a little fuzzy since it does not explicitly say persisting between what.

**Comment by sbc100** [comment]
> > Is the CI error the last issue here?
> 
> This is not an issue of this PR. I added the tests, so we know that environment variables are not set on Windows (the behavior is broken). We need to fix the tests in other pull requests such as in #91.
> 

Can you remove the tests from this PR then?   somehow disable them or mark them as currently known/expected to fail?  We don't want to land a change with the CI being red.

File: scripts/test_permanent.ps1 [hunk 5]
```diff
@@ -0,0 +1,67 @@
+$cwd=(Get-Location).Path
+
+try {
+
+    cd([System.IO.Path]::GetDirectoryName((resolve-path "$PSScriptRoot")))
+
+    ./emsdk.ps1 install latest
+    ./emsdk.ps1 activate latest --permanently
```
[the remark this answers is not in the document]
Comment (aminya) [reply scripts/test_permanent.ps1]: Fixed in https://github.com/emscripten-core/emsdk/pull/621/commits/405cf3baba48f2d49544d2b143522673f853b4e9

**Comment by aminya** [comment]
> > > Is the CI error the last issue here?
> > 
> > 
> > This is not an issue of this PR. I added the tests, so we know that environment variables are not set on Windows (the behavior is broken). We need to fix the tests in other pull requests such as in #91.
> 
> Can you remove the tests from this PR then? somehow disable them or mark them as currently known/expected to fail? We don't want to land a change with the CI being red.

I caught the failures in https://github.com/emscripten-core/emsdk/pull/621/commits/f2575062c60e438a75e3e3ecc7efa7d28e56e6c5

This PR is ready to go.

**Comment by kripken** [comment]
I don't have a preference between `--persistent` vs `--permanent `.

I think if the code looks good to you @sbc100 , let's land it!

File: emsdk.py [hunk 6]
```diff
@@ -2971,8 +2973,8 @@ def print_tools(t):
     fetch_emscripten_tags()
     return 0
   elif cmd == 'activate':
-    if arg_global:
-      print('Registering active Emscripten environment globally for all users.')
+    if arg_permanent:
+      print('Registering active Emscripten environment premenantly')
```
Comment (sbc100) [inline emsdk.py:2977]: sp: permanently

File: emsdk.py [hunk 7]
```diff
@@ -2775,7 +2770,14 @@ def extract_bool_arg(name):
 
   arg_old = extract_bool_arg('--old')
   arg_uses = extract_bool_arg('--uses')
+  arg_permanent = extract_bool_arg('--permanent')
   arg_global = extract_bool_arg('--global')
+  if arg_global:
+    print('--global is deprecated. Use `--system` to set the environment variables for all users')
+    arg_permanent = True
```
Comment (sbc100) [inline emsdk.py:2777]: I think the old behaviour was the "system" behaviour so this should probably be `arg_system = True` and `arg_permenant = True` and happen after the arg_system handler?

File: emsdk.py [hunk 8]
```diff
@@ -2737,15 +2727,20 @@ def main():
...
                                 - Activates the given tool or SDK in the
-                                  environment of the current shell. If the
-                                  --global option is passed, the registration
-                                  is done globally to all users in the system
-                                  environment.  If a custom compiler version was
-                                  used to override the compiler to use, pass
-                                  the same --vs2017/--vs2019 parameter
+                                  environment of the current shell.
+
+                                - If the `--permanent` option is passed, the the environment
```
Comment (sbc100) [inline emsdk.py:2735]: the the -> then the

File: emsdk.py [hunk 9]
```diff
@@ -2737,15 +2727,20 @@ def main():
...
+                                  variables are set permanently for the current user.
+
+                                - If the `--system` option is passed, the registration
+                                  is done for all users of the system
+                                  (uses Machine environment variables).
```
Comment (sbc100) [inline emsdk.py:2740]: Should we mention this requires admin privs here?

**Review by sbc100 — APPROVED** [review]
lgtm with a couple more comments.   Ready to land once they are addressed.

File: emsdk.py (hunk 6 above)
Comment (aminya) [inline emsdk.py:2977]:
```suggestion
      print('Registering active Emscripten environment permanently')
```

File: emsdk.py (hunk 8 above)
Comment (aminya) [inline emsdk.py:2735]:
```suggestion
                                - If the `--permanent` option is passed, then the environment
```

File: emsdk.py [hunk 10]
```diff
@@ -2737,15 +2727,20 @@ def main():
...
-                                  used to override the compiler to use, pass
-                                  the same --vs2017/--vs2019 parameter
+                                  environment of the current shell.
+
+                                - If the `--permanent` option is passed, then the environment
+                                  variables are set permanently for the current user.
+
+                                - If the `--system` option is passed, the registration
+                                  is done for all users of the system
+                                  (uses Machine environment variables).
```
Comment (aminya) [inline emsdk.py:2740]:
```suggestion
                                  is done for all users of the system. 
                                  This needs admin privileges
                                  (uses Machine environment variables).
```

File: emsdk.py [hunk 11]
```diff
@@ -2775,7 +2771,14 @@ def extract_bool_arg(name):
 
   arg_old = extract_bool_arg('--old')
   arg_uses = extract_bool_arg('--uses')
+  arg_permanent = extract_bool_arg('--permanent')
   arg_global = extract_bool_arg('--global')
+  if arg_global:
+    print('--global is deprecated. Use `--system` to set the environment variables for all users')
+    arg_permanent = True
```
Comment (aminya) [inline emsdk.py:2778]:
```suggestion
    arg_system = True
```

File: emsdk.py (hunk 7 above)
Comment (aminya) [reply emsdk.py]: https://github.com/emscripten-core/emsdk/pull/621/commits/0f85911cfad108444f97c4e40402a592845fc07e

File: emsdk.py (hunk 3 above)
Comment (sbc100) [reply emsdk.py:2777]: Sorry I think this still doesn't work becaue arg_system is assigned on the line below.

Comment (sbc100) [reply emsdk.py:2777]: I think if you just move it below it should work.

Comment (aminya) [reply emsdk.py:2778]: It is fixed now.

**Comment by sbc100** [comment]
Looks like the removal of the second call to causes the issue:

```
current_user_path = win_get_environment_variable('PATH', system=False)
```

Previously it was adding the user and system PATHs together I think.

I guess we should probably just ignore the registry when running in normal (non-persistent) mode and just look at the current os.environ['PATH']?

**Pull request merged**